internal/controller/controllers/hypershiftagentserviceconfig_controller.go

```diff
@@ -128,8 +122,44 @@ func (hr *HypershiftAgentServiceConfigReconciler) Reconcile(origCtx context.Cont
 		return ctrl.Result{}, err
 	}
 
+	// Initialize ASC to reconcile components according to instance context
+	asc.initHASC(hr, instance)
+
+	// Reconcile hub components
+	if result, err := hr.reconcileHubComponents(ctx, log, asc); err != nil {
+		return result, err
+	}
+
+	// Ensure instance namespace exists on spoke cluster
+	if err := hr.ensureSpokeNamespace(ctx, spokeClient, instance.Namespace); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Switch to spoke client in the reconciler context.
+	// We need to store the hub client as ASC stores a reference to AgentServiceConfigReconcileContext.
+	hubClient := asc.rec.Client
+	defer func() {
+		asc.rec.Client = hubClient
```

Comment:
Why does this matter? `asc` is a local variable scoped to this `Reconcile` function call. Why would we care about the state after this call?

Reply:
asc is indeed a local env, but we change Client in rec - which is a pointer to AgentServiceConfigReconcileContext in the reconciler.

Reply:
Then this would not be changed because reconcile can be called in parallel

Reply:
Do we really need to worry about this if we're using the default 1 concurrent Reconciles?
https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/controller/controller.go#L39

Reply:
ok, created a new instance of AgentServiceConfigReconcileContext in initHASC, so it should mitigate this issue.

Reply:
Yes. the defensive copy should be a good solution path